### DISCUSSION TITLE
What does app.set('socketio', io) do?

### DISCUSSION THREAD

**Discussion by oshihirii** [socketio/socket.io#4157] [Q&A] [answered]
**Environment**

socket.io:  4.3.1  
socket.io-client:  4.3.2  
express:  4.17.1
node:  v14.18.1  

After a `socket.io` upgrade, I am revisiting my setup and wanting to make sure I understand my implementation:

I use this:  

```js
import express from "express";
const app = express();
const http = require('http');
const server = http.createServer(app);
const io = require('socket.io')(server);
app.set('socketio', io);   // <--------------- what does this do?  
```

Whereas this guide:

https://socket.io/get-started/chat#integrating-socketio

Uses this:

```js
const express = require('express');
const app = express();
const http = require('http');
const server = http.createServer(app);
const { Server } = require("socket.io");
const io = new Server(server);
```

And this guide:

https://github.com/socketio/socket.io#in-conjunction-with-express

Uses this:

```js
const app = require('express')();
const server = require('http').createServer(app);
const io = require('socket.io')(server);
```

**My question is:**

What does `app.set('socketio', io)` do?  

I can see it is not used in any of the other guides mentioned above (maybe it was in older versions?).  

I've looked at the docs, but I don't understand what it means:

http://expressjs.com/en/api.html#app.set

And there is some related discussion here:

https://stackoverflow.com/a/31277123  

**Reference**

For reference, my `.babelrc.json` is:

```js
{
    "presets": [
        [
            "@babel/preset-env",
            {
                "targets":
                {
                    "node": "current"
                }
            }
        ]
    ]
}
```

And I run it on my server side file from `package.json` with:

```js
babel app_es6.js --out-file app.js
```

**Comment by darrachequesne** [comment, marked as answer]
`app.set('socketio', io)` stores the Socket.IO server so you can retrieve it later with `app.get('socketio')`. For example:

```js
app.post('/hello', (req, res) => {
  app.get('socketio').emit('hello');
  res.send('done');
})
```

**Comment by oshihirii** [comment]
Thank you for answer, unfortunately I am still trying to understand.

Does "retrieve it later" mean "retrieve it outside of the `on connection`  function", eg:  
 
```js

import express from "express";
const app = express();
const http = require('http');
const server = http.createServer(app);
const io = require('socket.io')(server);
app.set('socketio', io); 

// io and socket work within the function...
io.on('connection', async (socket) => {    

// emit data  
socket.emit('heres_your_socket_id', socket.id);

// listen for data 
socket.on('some_event', async (data) => {  // do things   } );  

// sanity check - get all connected sockets  
const socket_ids_set = await io.allSockets();
console.log("socket_ids_set is currently:");
console.log(socket_ids_set ); 

// join a room
socket.join(room);

// another sanity check - get all connected sockets within a room  
const room_socket_ids_set = await io.of("/").in(room).allSockets();
console.log(`room_socket_ids_set in "${room}" is:`);
console.log(room_socket_ids_set);

});

/* 

... but 'io' and 'socket' do not work outside of the function? 

eg you can't do this here:    

socket.emit()  
io.emit()  

therefore we need to use this outside the function?  

app.get('socketio').emit('hello');

*/
```

Why can't we use `io` outside of the function if it has been defined as a global variable (via `const io = require('socket.io')(server)`)?

**Reply by darrachequesne** [reply]
> Why can't we use io outside of the function if it has been defined as a global variable (via const io = require('socket.io')(server))?

You totally can! Actually, I can't think of a case where using `app.(s|g)et('socketio')` is needed.
